lms/views/api/sync.py

```diff
@@ -0,0 +1,43 @@
+import hashlib
+
+from pyramid.view import view_config
+
+from lms.models import HGroup, h_group_name
+
+
+@view_config(
+    route_name="api.sync", request_method="POST", renderer="json", permission="sync_api"
+)
+def sync(request):
+    authority = request.registry.settings["h_authority"]
+    canvas_api_svc = request.find_service(name="canvas_api_client")
+    lti_h_svc = request.find_service(name="lti_h")
+
+    context_id = request.json["course"]["context_id"]
+    custom_canvas_course_id = request.json["course"]["custom_canvas_course_id"]
+    tool_consumer_instance_guid = request.json["lms"]["tool_consumer_instance_guid"]
+
+    if request.lti_user.is_learner:
+        # For learners we only want the client to show the sections that the
+        # student belongs to, so fetch only the user's sections.
+        sections = canvas_api_svc.authenticated_users_sections(custom_canvas_course_id)
+    else:
+        # For non-learners (e.g. instructors, teaching assistants) we want the
+        # client to show all of the course's sections.
+        sections = canvas_api_svc.course_sections(custom_canvas_course_id)
+
+    groups = []
+
+    for section in sections:
+        # Generate a unique authority_provided_id for the section group.
+        hash_object = hashlib.sha1()
+        hash_object.update(tool_consumer_instance_guid.encode())
+        hash_object.update(context_id.encode())
+        hash_object.update(str(section["id"]).encode())
+        authority_provided_id = hash_object.hexdigest()
```

Comment:
This sticks out as kind of obscure compared to the simplicity of the rest. I think readability would be improved by popping this out to a separate function.

On a related note, we've had a rash of functions very like this. I wonder if we are missing something in models like:

```python
def hashed_id(*parts):
    hash_object = hashlib.sha1()
    
    for part in parts:
        hash_object.update(str(part).encode())

    return hash_object.hexdigest()
    

# ... meanwhile

from models.? import hashed_id


def _section_id(instance_guid, context_id, section):
    # At this point should we even bother?
    return hashed_id(instance_guid, context_id, section['id'])
```

Reply:
I think generating the `HGroup` for a section is quite specific to this view (it depends on the view's `tool_consumer_instance_guid` and `context_id` params, as well as on the section dicts retrieved from `canvas_api_svc`) so I don't think it bears extracting out into a service or anything. So I think we're just talking about a collaborator function for the view, as in your example.

I hate to say it but I think an inner function is nicer here due to the ability to use closures so you don't have to pass long arguments. Here it is with an outer helper function:

```python
    groups = [
        _group(tool_consumer_instance_guid, context_id, section) for section in sections
    ]

    lti_h_svc.sync(groups)

    return [group.groupid(authority) for group in groups]


def _group(tool_consumer_instance_guid, context_id, section):
    hash_object = hashlib.sha1()
    hash_object.update(tool_consumer_instance_guid.encode())
    hash_object.update(context_id.encode())
    hash_object.update(str(section["id"]).encode())
    authority_provided_id = hash_object.hexdigest()

    return HGroup(h_group_name(section["name"]), authority_provided_id,)

```

And here's the inner-function version, which I think reads nicer:

```python
    def authority_provided_id(section):
        hash_object = hashlib.sha1()
        hash_object.update(tool_consumer_instance_guid.encode())
        hash_object.update(context_id.encode())
        hash_object.update(str(section["id"]).encode())
        return hash_object.hexdigest()

    groups = [
        HGroup(h_group_name(section["name"]), authority_provided_id(section),)
        for section in sections
    ]

    lti_h_svc.sync(groups)

    return [group.groupid(authority) for group in groups]
```

Reply:
I think a `models.hashed_id()` function might be a good idea but I'd like to leave that for a separate PR

Reply:
https://github.com/hypothesis/lms/pull/1680/commits/0529e0a550d086b45b9a57a270235ed640bfc68a